### ISSUE TITLE
When to destroy activity's presenter

### ISSUE THREAD

**Issue by konmik** [konmik/nucleus#82]
We need a more reliable way to detect when to destroy activity's presenter.

Here is where the issue that started the discussion: https://github.com/konmik/nucleus/issues/42#issuecomment-204233175

TL/DR: activity's `finish()` can be called some time after `onPause`, so this code will leak the activity's  presenter: https://github.com/konmik/nucleus/blob/master/nucleus/src/main/java/nucleus/view/NucleusActivity.java#L76

Possible solutions:

a) Leverage retained fragments. This will solve the presenter destruction issue but we will leak background tasks sometimes.

b) Extend `PresenterLifecycleDelegate` lifecycle to explicitly call `onDestroy`.

c) Install system-wide lifecycle listener `registerActivityLifecycleCallbacks`.

**Comment by duo2005duo** [comment]
if a activity may not restart forever,why not leak the background tasks?

**Comment by konmik** [comment]
Because some background tasks can be very resource-consuming (streaming, image processing, etc).

**Comment by duo2005duo** [comment]
I think the system will drop an activity only after a certain time(
maybe several minutes).The resource-consuming tasks will be completed in the certain time and the result have already been saved with the help of `saveInstanceState`.
 if the tasks is long enough to exceed the certain time.Just let it be a tasks with no end.Tasks will accumulate in such situation,which waste the system resource and the system will crash in the end.

**Comment by konmik** [comment]
Android never drops activities after some time. It drops them only with the entire process.

**Comment by duo2005duo** [comment]
No,If an activity is paused or stopped, the system can drop the activity from memory by either asking it to finish, or simply killing its process

**Comment by konmik** [comment]
Do you have any references to docs or tests?

I created a test once which runs 100500 activities one on top of another - just to check if Android will finish some of background Activities when it becames short on memory. It didn't.

**Comment by konmik** [comment]
There is also a discussion on stackoverflow somewhere with the same conclusion.

**Comment by duo2005duo** [comment]
Yes, [doc](http://developer.android.com/reference/android/app/Activity.html#ActivityLifecycle) I think the system drop the activity in this situation because of  time but not memory.

**Comment by konmik** [comment]
:D such an obvious lie in official docs. I never had a sutuation when Android finishes my Activities without my own command. It can finish them indirectly if I say FLAG_ACTIVITY_CLEAR**, but it never finishes them by itself.

How much time do you think we need for activity to be finished because of time?

**Comment by konmik** [comment]
I've found something interesting: http://developer.android.com/guide/components/tasks-and-back-stack.html

"If the user leaves a task for a long time, the system clears the task of all activities except the root activity. "

**Comment by konmik** [comment]
http://developer.android.com/guide/topics/manifest/activity-element.html#always

**Comment by duo2005duo** [comment]
Ok,It said 30 minutes .I happened to see it before and I forget where,you find it.

**Comment by konmik** [comment]
It is always nice to know more info about these lifecycles, thanks for mentioning them! :)

**Comment by bejibx** [comment]
Hey, did you red [this article](https://medium.com/@czyrux/presenter-surviving-orientation-changes-with-loaders-6da6d86ffbbf#.srm8i0bqc)? i think it is really nice way to bound to Activity lifecycle.

**Comment by konmik** [comment]
@bejibx I hope you're kidding. This ugly workaround using one of the most buggy Android APIs will never make its way to my code. Never.

**Comment by bejibx** [comment]
Workaround is kind of ugly, true, but it seems to me pretty much every possible solution to beat Activity lifecycle is ugly. Still I'm a bit confused, because loaders seems to work like a charm in my projects before.

**Comment by konmik** [comment]
@bejibx you're lucky then. Their lifecycle is not the same as Activity has and not the same as process has. So, they obviously add complexity instead of reducing it. And, in addition to the added complexity, it has so many callbacks and unneeded API complexities like sending arguments in a Bundle, forceLoad... they just suck.

**Comment by konmik** [comment]
Nucleus exists only because loaders do not work.

**Comment by bejibx** [comment]
I got your point and I'm definitely agree with you about Loaders API over-complexity. I hope I'll have some time on this weekends to investigate Android Activity and Loader sources to see if system somehow diferentiate restart from complete destroy.

**Comment by konmik** [comment]
https://github.com/konmik/nucleus/commit/44c7722e32d1d54eb95f126586c79f114c4b14a0

**Comment by duo2005duo** [comment]
44c7722 do not fix the situaction in case 2 "restart-activity" in your post.In that situation,the method `isFinishing` still returns false.

**Comment by konmik** [comment]
@duo2005duo Which post do you mean?

**Comment by konmik** [comment]
This one?

"Case 2: An Activity restart happens when a user has set "Don't keep activities" checkbox in Developer's settings and another activity becomes topmost." ?

**Comment by duo2005duo** [comment]
@konmik  Yes

**Comment by konmik** [comment]
I think it will have `isFinishing` during its next `onDestroy`.

**Comment by bejibx** [comment]
@konmik this is true, just checked it in sample project.

**Comment by duo2005duo** [comment]
Of course.But do you think the retained fragment lifecycle is similar to current presenter lifecycle？Combining staticholder and saveInstanceState just do the same with retained fragment.I think it voliate KISS.

**Comment by duo2005duo** [comment]
And I think the loader lifecycle is the same with retain fragment.They are both `NonConfigurationInstances`.

**Comment by konmik** [comment]
@duo2005duo the idea is to not have the same background task running twice. Without having a static variable we can't detect if the task is still running. Ratained fragments do not fit because they get destroyed during "case 2".

**Comment by duo2005duo** [comment]
Yes,But I think case 2 is not so important,The user will not do this except the developer,so I said "similar". And it still need to be tested whether it will have isFinishing during its next onDestroy for cleaning task  we talked about yesterday

**Comment by duo2005duo** [comment]
And I think it will not have its fininishing,the activity will not destroy twice without have a second onCreate.

**Comment by konmik** [comment]
@duo2005duo You're right this is not *very* important. But why complicate things and use fragments if we can do the same without them? Fragments is not the only thing we need presenters for.

It will have a second `onCreate` and a second `onDestroy` with `isFinishing`. I tested.

**Comment by duo2005duo** [comment]
The result is interesting !:-D

**Comment by bejibx** [comment]
Also I found this mentioned earlier [stackoverflow conversation](http://stackoverflow.com/a/7576275/1315647) about Android destroying activities. So I think when lead Android engineer says

> The only memory management that impacts activity lifecycle is the global memory across all processes, as Android decides that it is running low on memory and so need to kill background processes to get some back.

we can trust him.

**Comment by bejibx** [comment]
I think I found this rare case when we can leak our presenters because ```isFinishing()``` returns ```false```.
What we need is "Do not keep Activities" developer option enabled and following backstack:
Activity 1 -> Activity 2 -> [Activity 3]
So at this point Activity 1 and 2 was destroyed with ```isFinishing() == false``` and we expect to destroy our pesenters when user navigates to them later with back button. Now imagine if we use [up navigation](http://developer.android.com/intl/ru/training/implementing-navigation/ancestral.html) from Activity 3 to Activity 1. ```onDestroy``` on Activity 2 will not be called (tested) and we'll lost our presenter.

**Comment by konmik** [comment]
@bejibx You're right, we will have a leak in this case. I was already thinking about this.

I think we need to recommend `android:alwaysRetainTaskState` flag to prevent such leaks.

**Comment by konmik** [comment]
While this flag doesn't help with "Do not keep Activities" it will help with Android finishing activities by itself.

**Comment by bejibx** [comment]
But did you test if ```isFinishing() == false``` when Android destroys Activities this way? I was trying to test it yesterday - I leaved my task for something around 3 hours and when I checked, it turned out that Android just killed my process. Maybe thats because my test was incorrect - I had only 1 task.

Also I tested what is gonna happen with flag ```android:clearTaskOnLaunch="true"```. Find out ```isFinishing() == true``` on all destroyed Activities.

**Comment by bejibx** [comment]
Also just found [this](http://stackoverflow.com/q/18847342/1315647) interresting stackoverflow article.

**Comment by bejibx** [comment]
```java
final ActivityRecord resetTaskIfNeededLocked(ActivityRecord taskTop,
            ActivityRecord newActivity) {
        boolean forceReset =
                (newActivity.info.flags & ActivityInfo.FLAG_CLEAR_TASK_ON_LAUNCH) != 0;
        if (ACTIVITY_INACTIVE_RESET_TIME > 0
                && taskTop.task.getInactiveDuration() > ACTIVITY_INACTIVE_RESET_TIME) {
            if ((newActivity.info.flags & ActivityInfo.FLAG_ALWAYS_RETAIN_TASK_STATE) == 0) {
                forceReset = true;
            }
        }
//...
```
If we look at [ActivityStack#resetTaskIfNeededLocked(ActivityRecord, ActivityRecord)](http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/com/android/server/am/ActivityStack.java#2389) function we can see that no matter why activity should be killed - because of inactive reset time came off or because ```clearTaskOnLaunch``` flag was setted on Activity, actual Activity-by-time-removal mechanism stays the same, it only affects ```forceReset``` flag. So I assume if ```isFinishing()``` correctly returns ```true``` in my second test, we'll have same result for Activities destroyed by inactive reset mechanism.

**Comment by konmik** [comment]
I'm thinking about retained fragments and `onRetainCustomNonConfigurationInstance` :(

**Comment by konmik** [comment]
Here is a possible implementation, BTW: https://github.com/square/flow/blob/master/flow/src/main/java/flow/InternalLifecycleIntegration.java

I need to test retained fragments over edge cases.

**Comment by bejibx** [comment]
Calm down for a moment, we just need to test first what happens with Activity when system destroys it because of time. [It is confirmed](http://stackoverflow.com/q/18847342/1315647) this could be easily reproduced on Android 2.3:
> This behaviour can be easily reproduced on devices running Gingerbread and earlier. Launch an app and create some back history, then hit the home button and wait half an hour. Launch the app again from the home screen and the state has been cleared as if it were starting a new task. Perfect.
I'll test this on emulator tonight or tomorrow if I there is no time today and report results.

**Comment by konmik** [comment]
@bejibx Even if it will call `onDestroy` with `isFinishing` this will not solve the "Do not keep activities" + "Clear stack" issue.

**Comment by konmik** [comment]
moving to retained fragments will also automatically solve fragment's presenter lifecycle problem where we need to manually destroy presenter.

**Comment by konmik** [comment]
We will have sometimes several instances of the same background task running but it will, at least, be a predictable scenario every other android developer is facing.

**Comment by duo2005duo** [comment]
@konmik  The RxJava will stop the unneccessary task instances as soon as possible with the help of `isUnscribed()`. So it's OK.

**Comment by bejibx** [comment]
Hey, can we use [isChangingConfigurations()](http://developer.android.com/reference/android/app/Activity.html#isChangingConfigurations()) instead of [isFinishing()](http://developer.android.com/reference/android/app/Activity.html#isFinishing())? So presenter will NOT be destroyed only if it configuration change incoming.

**Comment by konmik** [comment]
@bejibx awesome. I will try it.

**Comment by konmik** [comment]
3.0.0-beta is on maven

**Comment by konmik** [comment]
https://github.com/konmik/nucleus/commit/2863995e40d65d144843f91e0e646a5d4be6248d
